src/tensor_products.jl

```diff
@@ -42,43 +42,44 @@ end
 const ⊡ = dcontract
 
 # Specialized methods for symmetric tensors
-@gen_code function dcontract{dim, T1, T2, M}(S1::SymmetricTensor{2, dim, T1, M}, S2::SymmetricTensor{2, dim, T2, M})
-    Tv = typeof(zero(T1) * zero(T2))
-    @code :($(Expr(:meta, :inline)))
-    @code :(s = zero($Tv);
-            data1 = get_data(S1);
-            data2 = get_data(S2))
-     for k in 1:M
-        if is_diagonal_index(dim, k)
-            @code :(@inbounds s += data1[$k] * data2[$k])
+@generated function dcontract{dim}(S1::SymmetricTensor{2, dim}, S2::SymmetricTensor{2, dim})
+    idx2(i,j) = compute_index(SymmetricTensor{2, dim}, i, j)
+    ex = Expr[]
+    for i in 1:dim, j in i:dim
+        if i == j
+            push!(ex, :(get_data(S1)[$(idx2(i, j))] * get_data(S2)[$(idx2(i, j))]))
         else
-            @code :(@inbounds s += 2 * data1[$k] * data2[$k])
+            push!(ex, :(2 * get_data(S1)[$(idx2(i, j))] * get_data(S2)[$(idx2(i, j))]))
         end
     end
-    @code :(return s)
+    exp = reduce((ex1,ex2) -> :(+($ex1, $ex2)), ex)
```

Comment:
`reduce` here seems to be the only difference from `StaticArrays` compared to what we did before we built on that